dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguagePluginHandler.java

```diff
@@ -195,36 +195,42 @@ public List<String> listFiles(String pathToDirectory) {
-            } else if (entry.getValue().genericFileType() == GenericFileType.TEST_PARAMETER_FILE) {
-                type = minimalLanguageInterface.getDescriptorLanguage().getTestParamType();
-            } else {
-                // For some languages this may be incorrect
-                type = FileType.DOCKERFILE;
-            }
-            if (minimalLanguageInterface.getDescriptorLanguage().isServiceLanguage()) {
-                // TODO: this needs to be more sophisticated
-                type = DescriptorLanguage.FileType.DOCKSTORE_SERVICE_YML;
+                // The language plugins don't necessarily set the file type
+                // so query the plugin to find out what the imported file
+                // type should be, because this is needed in downstream code.
+                // We assume if imported files are not descriptors or not test files they are Dockerfiles,
+                // however this may not be true for some languages, and we may have to change this
+                DescriptorLanguage.FileType type;
+                if (entry.getValue().genericFileType() == GenericFileType.IMPORTED_DESCRIPTOR) {
+                    type = minimalLanguageInterface.getDescriptorLanguage().getFileType();
+                } else if (entry.getValue().genericFileType() == GenericFileType.TEST_PARAMETER_FILE) {
+                    type = minimalLanguageInterface.getDescriptorLanguage().getTestParamType();
+                } else {
+                    // For some languages this may be incorrect
+                    type = FileType.DOCKERFILE;
+                }
+                if (minimalLanguageInterface.getDescriptorLanguage().isServiceLanguage()) {
+                    // TODO: this needs to be more sophisticated
+                    type = DescriptorLanguage.FileType.DOCKSTORE_SERVICE_YML;
+                }
+                String path = entry.getKey();
+                String fileContent = entry.getValue().content();
+                SourceFile sourceFile = SourceFile.limitedBuilder().type(type).content(fileContent).paths(path).build();
+                sourceFile.getMetadata().setTypeVersion(entry.getValue().languageVersion());
+                results.put(entry.getKey(), sourceFile);
             }
-            String path = entry.getKey();
-            String fileContent = entry.getValue().content();
-            SourceFile sourceFile = SourceFile.limitedBuilder().type(type).content(fileContent).paths(path).build();
-            sourceFile.getMetadata().setTypeVersion(entry.getValue().languageVersion());
-            results.put(entry.getKey(), sourceFile);
+            return results;
+        } catch (NullPointerException e) {
+            // ignore, cannot assume plugins are well-behaved and a badly-behaved plugin shouldn't bring the rest of the webservice down
+            LOG.error("plugin threw NullPointer exception, dodging");
+            return new HashMap<>();
```

Comment:
If a plugin has a fatal error, should we throw a `CustomWebApplicationException`?  Might be better if the registration process fails, rather than the import step "succeeding" but returning nothing.

Reply:
See above, partial results could be useful though. 
As above, I was able to view the primary but not imports by catching.

Reply:
This is probably ok for now, but in production, IMHO, it's bad for the plugin to fail and result in a "partial" registration.